src/socketio.quiver.js

```diff
@@ -263,23 +300,59 @@ QuiverSocialProvider.prototype.login = function(loginOpts, continuation) {
             this.connectAsClient_.bind(this, friend, null)).
                 catch(onConnectionFailure.bind(this, friend)));
       }
-      Promise.all(connectionPromises).then(finishLogin);
+      return Promise.all(connectionPromises);
     }.bind(this);
 
     this.connectLoop_(this.configuration_.self.servers,
-        this.connectAsOwner_.bind(this)).then(connectToFriends, connectToFriends);
+        this.connectAsOwner_.bind(this)).then(function(results) {
+      if (results.succeeded.length > 0) {
+        // If at least one connection succeeded, then we do have a working
+        // network, so any connection failures are because the server is in fact
+        // failing or unreachable.  Remove it from the active list.
+        // It will then be replaced by a friend's server.
+        results.failed.forEach(this.removeServer_, this);
+      }
+      return connectToFriends();
+    }.bind(this)).then(function() {
+      var deficit = QuiverSocialProvider.MAX_CONNECTIONS_ - this.countOwnerConnections_();
+      if (deficit > 0) {
+        /** @type {!Object<string, QuiverSocialProvider.server_>} */
+        var unusedDefaultServers = {};
+        QuiverSocialProvider.DEFAULT_SERVERS_.forEach(function(server) {
+          var key = QuiverSocialProvider.serverKey_(server);
+          if (!(key in this.connections_)) {
+            unusedDefaultServers[key] = server;
+          }
+        }, this);
+        return this.connectLoop_(unusedDefaultServers,
+            this.connectAsOwner_.bind(this), deficit);
+      }
+    }.bind(this)).then(finishLogin);
   }.bind(this));
 };
 
 /**
+ * @private @typedef {{
+ *   succeeded: !Array<!QuiverSocialProvider.server_>,
+ *   failed: !Array<!QuiverSocialProvider.server_>
+ * }}
+ */
+QuiverSocialProvider.connectLoopResults_ = undefined;
+
+/**
  * Async for-loop.  Try to connect to each server, and stop once we hit the
  * limit or run out of servers.
  * @param {!Object.<string, !QuiverSocialProvider.server_>} servers
  * @param {function(!QuiverSocialProvider.server_, Function)} connector
- * @return {!Promise}
+ * @param {number=} opt_limit Optional limit, defaults to MAX_CONNECTIONS
+ * @return {!Promise<!QuiverSocialProvider.connectLoopResults_>} Always fulfills.
+ *     
  * @private
  */
-QuiverSocialProvider.prototype.connectLoop_ = function(servers, connector) {
+QuiverSocialProvider.prototype.connectLoop_ = function(servers, connector,
```

Comment:
Currently connectLoop_ waits for each server to succeed or fail before continuing.  This means that if a server times out rather than immediately succeeds/rejects then it may take 30+ seconds to fail (per server).  This is especially bad if the timeouts occur during login - I don't think we want our UI to wait 30+ seconds for login to succeed/fail (currently login only fulfills after we have a success or failure from every connectAsClient_ and connectAsOwner_ call).

Some ideas:
* Can we have login fulfill as soon as we are connected to at least 1 owner server?  Then we can connect to the client servers later async without delaying the uProxy UI from advancing
* Maybe connectLoop could perform ~10 connections in parallel?  I'm not sure what the magic number is here, I get that if there are 1000 servers we don't want to make that many connections in parallel, but doing 1 at a time seems too slow.